File: scripts/data_paths.py
from __future__ import annotations

from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parent.parent
DATA_DIR = ROOT_DIR / "data"
MARKETS_DIR = DATA_DIR / "markets"
A_SHARE_DIR = MARKETS_DIR / "a_share"
SYSTEM_DIR = DATA_DIR / "system"
RESEARCH_REPORTS_DIR = DATA_DIR / "research_reports"
A_SHARE_REPORTS_DIR = RESEARCH_REPORTS_DIR / "a_share"

SYSTEM_FILES = {"sync_status.json", "auth.db"}
REPORT_MANIFEST_FILES = {"research_reports_manifest.json", "research_reports/a_share/manifest.json"}


def resolve_write_path(file_name: str, kind: str = "auto") -> Path:
    normalized = str(file_name).replace("\\", "/").lstrip("/")

    if kind == "system" or (kind == "auto" and normalized in SYSTEM_FILES):
        return SYSTEM_DIR / normalized

    if kind == "research_manifest" or normalized in REPORT_MANIFEST_FILES:
        return A_SHARE_REPORTS_DIR / "manifest.json"

    if kind == "research_dir":
        return A_SHARE_REPORTS_DIR / normalized

    if normalized.startswith(("markets/", "system/", "research_reports/")):
        return DATA_DIR / normalized

    return A_SHARE_DIR / normalized


def resolve_read_candidates(file_name: str, kind: str = "auto") -> list[Path]:
    normalized = str(file_name).replace("\\", "/").lstrip("/")
    primary = resolve_write_path(normalized, kind=kind)

    if kind == "research_manifest" or normalized in REPORT_MANIFEST_FILES:
        return [A_SHARE_REPORTS_DIR / "manifest.json", DATA_DIR / "research_reports_manifest.json"]

    if kind == "system" or (kind == "auto" and normalized in SYSTEM_FILES):
        return [SYSTEM_DIR / normalized, DATA_DIR / normalized]

    if normalized.startswith(("markets/", "system/", "research_reports/")):
        return [DATA_DIR / normalized]

    legacy = DATA_DIR / normalized
    if primary == legacy:
        return [primary]

    return [primary, legacy]


def resolve_existing_path(file_name: str, kind: str = "auto") -> Path:
    candidates = resolve_read_candidates(file_name, kind=kind)
    for candidate in candidates:
        if candidate.exists():
            return candidate
    return candidates[0]

File: scripts/dataPaths.js
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export const ROOT_DIR = path.resolve(__dirname, '..');
export const DATA_DIR = path.join(ROOT_DIR, 'data');
export const MARKETS_DIR = path.join(DATA_DIR, 'markets');
export const A_SHARE_DIR = path.join(MARKETS_DIR, 'a_share');
export const SYSTEM_DIR = path.join(DATA_DIR, 'system');
export const RESEARCH_REPORTS_DIR = path.join(DATA_DIR, 'research_reports');
export const A_SHARE_REPORTS_DIR = path.join(RESEARCH_REPORTS_DIR, 'a_share');

const SYSTEM_FILES = new Set(['sync_status.json', 'auth.db']);

const toPosix = (value) => String(value || '').replace(/\\/g, '/').replace(/^\/+/, '');

const buildCandidates = (fileName, kind = 'auto') => {
  const normalized = toPosix(fileName);
  if (!normalized) return [];

  if (kind === 'system' || (kind === 'auto' && SYSTEM_FILES.has(normalized))) {
    return [path.join(SYSTEM_DIR, normalized), path.join(DATA_DIR, normalized)];
  }

  if (kind === 'research-manifest' || normalized === 'research_reports_manifest.json' || normalized === 'research_reports/a_share/manifest.json') {
    return [
      path.join(A_SHARE_REPORTS_DIR, 'manifest.json'),
      path.join(DATA_DIR, 'research_reports_manifest.json'),
    ];
  }

  if (kind === 'research-dir') {
    return [path.join(A_SHARE_REPORTS_DIR, normalized)];
  }

  if (normalized.startsWith('markets/') || normalized.startsWith('system/') || normalized.startsWith('research_reports/')) {
    return [path.join(DATA_DIR, normalized)];
  }

  return [path.join(A_SHARE_DIR, normalized), path.join(DATA_DIR, normalized)];
};

export const resolveDataWritePath = (fileName, kind = 'auto') => buildCandidates(fileName, kind)[0];

export const resolveDataReadCandidates = (fileName, kind = 'auto') => buildCandidates(fileName, kind);

export const resolveExistingDataPath = async (fileName, kind = 'auto') => {
  const candidates = buildCandidates(fileName, kind);
  for (const candidate of candidates) {
    try {
      await fs.access(candidate);
      return candidate;
    } catch {
      // ignore
    }
  }
  return candidates[0] ?? null;
};

export const ensureParentDir = async (filePath) => {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
};
